packages/react/src/shared/proxy/utilsProxy.ts

```diff
@@ -146,6 +146,63 @@ type DecorateProcedure<
   ): InfiniteData<inferProcedureOutput<TProcedure>> | undefined;
 };
 
+/**
+ * This function will traverse a type that has been constructed with a
+ * { __Router:{ [procedureOrRouterName]:({Type We Want}) | {__Router:{...}} }
+ * structure and extract all wanted types as a flat type union
+ *
+ * This is the only way I have found of bringing all types across a router to
+ * top level for us to merge them together for users to use!
+ */
+type ExtractUnionTypesFromRouterNest<RouterNest> = RouterNest extends {
+  __Router: infer ProceduresOrRouters;
+}
+  ? {
+      [Key in keyof ProceduresOrRouters]:
+        | (ProceduresOrRouters[Key] extends {
+            __Router: any;
+          }
+            ? never
+            : ProceduresOrRouters[Key]) // If it is not a router, extract the type we wanted and add to the union
+        | ExtractUnionTypesFromRouterNest<ProceduresOrRouters[Key]>; // Pass it to recursion regardless incase it is a router
+    }[keyof ProceduresOrRouters] // This also omits never keys 👍
+  : never;
+
+type RecursivelyInferAllQueryInputTypesUnderRouter<TRouter extends AnyRouter> =
+  {
+    __Router: /** Decorate as a router so we can safely recurse to flatten later */ {
+      [TKey in keyof TRouter['_def']['record']]: TRouter['_def']['record'][TKey] extends infer SubProcedureOrRouter
+        ? SubProcedureOrRouter extends QueryProcedure<any>
+          ? inferProcedureInput<SubProcedureOrRouter> extends void
+            ? never
+            : inferProcedureInput<SubProcedureOrRouter>
+          : SubProcedureOrRouter extends AnyRouter
+          ? RecursivelyInferAllQueryInputTypesUnderRouter<SubProcedureOrRouter>
+          : never
+        : never;
+    };
+  };
+
+type InferAllRouterQueryInputTypes<TRouter extends AnyRouter> =
+  ExtractUnionTypesFromRouterNest<
+    RecursivelyInferAllQueryInputTypesUnderRouter<TRouter>
+  >;
```

Comment:
Awesome work on these types! I'm unsure about the recursion, seems like it could get messy, but here's an alternative type:
```ts
type InferAllRouterQueryInputTypes<TRouter extends AnyRouter> = {
  [TKey in keyof TRouter['_def']['record']]: TRouter['_def']['record'][TKey] extends infer ProcedureOrRouter
    ? ProcedureOrRouter extends AnyQueryProcedure
      ? ProcedureArgs<TRouter['_def']['record'][TKey]>[0]
      : ProcedureOrRouter extends AnyRouter
      ? InferAllRouterQueryInputTypes<ProcedureOrRouter>
      : never
    : never;
} & DecorateRouterProcedure<TRouter>;
```

Reply:
Maybe we can compare build times on https://github.com/trpc/trpc/tree/next/examples/next-big-router on this branch vs `next`

Reply:
Fantastic! Thanks @sachinraja !

I couldn't quite get the exact syntax you wrote above to work _(Possibly `AnyQuery` is in new type changes I haven't merged yet and it didn't like the types of `ProcedureArgs<ProcedureOrRouter['_def']>[0]`)_  but with some tweaking taking your intent this has the same result: 

```ts
/**
 * A type that will traverse all procedures and sub routers of a given router to create a union of
 * their possible input types
 */
type InferAllRouterQueryInputTypes<TRouter extends AnyRouter> = {
  [TKey in keyof TRouter['_def']['record']]: TRouter['_def']['record'][TKey] extends infer ProcedureOrRouter
    ? ProcedureOrRouter extends QueryProcedure<any>
      ? inferProcedureInput<ProcedureOrRouter>
      : ProcedureOrRouter extends AnyRouter
      ? InferAllRouterQueryInputTypes<ProcedureOrRouter> // Recurse as we have a sub router!
      : never
    : never;
}[keyof TRouter['_def']['record']]; // This flattens results into a big union and ignores never keys
```

This is an awful lot more elegant as it removes the need for my Router nesting business, I had to resort to 👍👍 However, it is doing the same traverse across all sub routers that I also had concerns about performance wise ... I think the new improved version is saving the type compiler doing it twice compared to my old solution though so it may be 2x faster at least! 🚀.

Performance wise I guess I'll give it a shot on the new `bigBoi` example and if OK we see how it goes with use? 

I know we were keen to have types on this rather then an `unknown` so this seems the way!!

Reply:
Great minds @KATT... (maybe being too generous on my mind here!)... I didn't see your comment before I posted but my brain went to the new big router example too!

Reply:
I'll merge in the turbo type changes from `next` then we can do this comparison and see how it goes!